src/client/Microsoft.Identity.Client/Extensibility/AbstractConfidentialClientAcquireTokenParameterBuilderExtension.cs

```diff
@@ -22,15 +24,48 @@ public static class AbstractConfidentialClientAcquireTokenParameterBuilderExtens
         /// <param name="onBeforeTokenRequestHandler">An async delegate which gets invoked just before MSAL makes a token request</param>
         /// <returns>The builder to chain other options to.</returns>
         public static AbstractAcquireTokenParameterBuilder<T> OnBeforeTokenRequest<T>(
-            this AbstractAcquireTokenParameterBuilder<T> builder, 
-            Func<OnBeforeTokenRequestData, Task> onBeforeTokenRequestHandler) 
+            this AbstractAcquireTokenParameterBuilder<T> builder,
+            Func<OnBeforeTokenRequestData, Task> onBeforeTokenRequestHandler)
             where T : AbstractAcquireTokenParameterBuilder<T>
-        {            
+        {
+            if (builder.CommonParameters.OnBeforeTokenRequestHandler != null && onBeforeTokenRequestHandler != null)
+            {
+                throw new InvalidOperationException("Cannot set OnBeforeTokenRequest handler twice.");
+            }
+
             builder.CommonParameters.OnBeforeTokenRequestHandler = onBeforeTokenRequestHandler;
 
             return builder;
         }
 
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="builder"></param>
+        /// <param name="addIn"></param>
+        /// <returns></returns>
+        public static AbstractAcquireTokenParameterBuilder<T> WithAddIn<T>( // TODO: bogavril - support a list of add-ins ? 
```

Comment:
Not sure a list is needed. The question will be "in what order do we apply the add ins?" Will customers be aware of what order to use? Seems like it can be a pain point down the road. Probably should limit it to only allowing one add in for now?